company.el

```diff
@@ -2034,10 +2048,27 @@ followed by `company-search-toggle-filtering'."
 (defun company-select-next (&optional arg)
   "Select the next candidate in the list.
 
-With ARG, move by that many elements."
+With ARG, move by that many elements.
+
+Possible state transitions:
+- (arg > 0) unselected -> first candidate selected
+- (arg < 0) first candidate selected -> default selection
+- (arg < 0 wrap-round) unselected -> last candidate selected
+- (arg < 0 no wrap-round) unselected -> unselected"
   (interactive "p")
   (when (company-manual-begin)
-    (company-set-selection (+ (or arg 1) company-selection))))
+    (let ((selection (if company-selection
+                         (+ (or arg 1) company-selection)
+                       (+ (or arg 1)
+                          (cond ((or (not arg) (> arg 0)) -1)
+                                (t 0))))))
+      ;; special case
```

Comment:
This indeed turned out to be pretty complex.

If we allowed `(arg = 1) last candidate selected -> default selection`, would it make things a bit more uniform and, as such, simpler?

Reply:
Sorry for dawdling on the review. In other situation I'd merge and then fiddle myself later, but here we have two users of the feature in question in the discussion. And I am not one of them.

Reply:
I tried to simplify this code so that the special handling of selected <-> unselected is handled in one function (`company--clamp-selection`) in a way that leaves `company-select-next` almost unmodified.
Please have a look and let me know if you like it better.
https://github.com/nikital/company-mode/commit/c19fa33076c1230761be5f44634b7bc11e1170e0

Reply:
Thank you @nikital. Sorry I didn't notice this comment earlier, I went to kind of a same approach to you and push a new commit. Allow forward selection to nil selection state does make thing easier.
@dgutov Can you help reviewing?

Reply:
Thank you both.

@nikital It's hard for me to directly compare both patches, but there are definitely things to like in yours (in particular, the generous commentary). If you believe @kiennq's latest version can be improved still, please submit a new PR on top of the new master, and we'll continue there. Thank you either way.

It's time to merge, let's get this in and collect user feedback, if any. Thanks to @kiennq for the work and patience.